apps/src/dance/ai/DanceAiModal.tsx

```diff
@@ -126,66 +186,105 @@ const DanceAiModal: React.FunctionComponent = () => {
+    setMode(Mode.EXPLANATION);
+  };
+
+  const handleUseClick = () => {
+    currentAiModalField?.setValue(
+      JSON.stringify({
+        inputs,
+        ...generatedEffects.current.goodEffect?.results,
+      })
+    );
+    onClose();
+  };
+
+  const handleLeaveExplanation = () => {
+    setMode(Mode.RESULTS);
+  };
+
+  // Animate through the generating process.
+  useInterval(
+    () => {
+      if (mode === Mode.GENERATING) {
+        // We do a deep copy into a new object to ensure that
+        // a re-render is triggered at the end of this work.
+        const currentGeneratingProgress: GeneratingProgress = {
+          ...generatingProgress,
+        };
+
+        if (currentGeneratingProgress.subStep < GENERATING_SUBSTEP_COUNT - 1) {
+          // Bump substep.
+          currentGeneratingProgress.subStep++;
+        } else if (
+          currentGeneratingProgress.step < BAD_GENERATED_RESULTS_COUNT
+        ) {
+          // Bump step and reset substep;
+          currentGeneratingProgress.step++;
+          currentGeneratingProgress.subStep = 0;
+        } else {
+          // Leave these values intact, and go to the results.
+          setMode(Mode.RESULTS);
+        }
+
+        // One-shot update of step & substep.
+        setGeneratingProgress(currentGeneratingProgress);
+      }
+    },
+    mode === Mode.GENERATING ? GENERATION_SUBSTEP_DURATION : undefined
+  );
+
   const startAi = async () => {
-    const responseJsonString = chooseEffects(inputs);
-    const result = JSON.parse(responseJsonString);
-
-    // "Pick" a subset of fields to be used.  Specifically, we exclude the
-    // explanation, since we don't want it becoming part of the code.
-    const pickedResult = (({
-      backgroundEffect,
-      backgroundColor,
-      foregroundEffect,
-    }) => ({
-      backgroundEffect,
-      backgroundColor,
-      foregroundEffect,
-    }))(result);
-
-    const fullResult = {inputs, ...pickedResult};
-
-    const fullResultJson = JSON.stringify(fullResult);
-
-    // The block value will be set to this JSON.
-    setResultJson(fullResultJson);
+    const result = chooseEffects(inputs, ChooseEffectsQuality.GOOD);
+    generatedEffects.current = {
+      badEffects: Array.from(Array(BAD_GENERATED_RESULTS_COUNT).keys()).map(
+        () => chooseEffects(inputs, ChooseEffectsQuality.BAD)
+      ),
+      goodEffect: chooseEffects(inputs, ChooseEffectsQuality.GOOD),
+    };
+  };
+
+  const getCurrentGeneratedEffect = () => {
+    if (generatingProgress.step < BAD_GENERATED_RESULTS_COUNT) {
+      return generatedEffects.current.badEffects[generatingProgress.step];
+    } else if (generatedEffects.current.goodEffect) {
+      return generatedEffects.current.goodEffect;
+    } else {
+      return undefined;
+    }
   };
 
   const handleConvertBlocks = () => {
     const blocksSvg = generateBlocksFromResult(
       Blockly.getMainWorkspace(),
-      resultJson
+      JSON.stringify(generatedEffects.current.goodEffect)
```

Comment:
nit: early return if `generatedEffects.current.goodEffect` is `undefined`?

Reply:
Good call; will add.  It shouldn't happen in reality, but good to be careful!